PWA installation prompt popup

Cerceis opened the pull request:
Added a pop-up to prompt user to install the PWA.

BaptisteD commented:
Hey ! 
When this popup is supposed to appears ? In which screen ? 
And will it appears again if the user say No ?

Cerceis commented:
> Hey !
> When this popup is supposed to appears ? In which screen ?
> And will it appears again if the user say No ?

It appears on the main index page. If the user check "Don't show again" then it will not pop up again. The same if the user already installed the PWA. 

As for compatibility, current its only working for Chrome and Edge. Does not work on Safari and Firefox. 
[compatibility](https://developer.mozilla.org/en-US/docs/Web/API/BeforeInstallPromptEvent)

BaptisteD commented:
Okay, thanks for the answer. 
Could you : 
- use inline text buttons rather that boxed ones and without icons. 
- Put the "close" button first and the "install" in second position (see attached image for reference) 
- 
![Screenshot_20220831-162623](https://user-images.githubusercontent.com/2566847/187703877-886ba24f-704b-42a9-987c-b37ffa925540.jpg)

- make this prompt appear on the list building page and not in the first page. Otherwise, you just land on the website to discover it and you're ask if you want to install it without having the time to at least check what it is (like the cookies prompt we suffer everywhere)

Cerceis commented:
> Okay, thanks for the answer. Could you :
> 
> * use inline text buttons rather that boxed ones and without icons.
> * Put the "close" button first and the "install" in second position (see attached image for reference)
> 
> ![Screenshot_20220831-162623](https://user-images.githubusercontent.com/2566847/187703877-886ba24f-704b-42a9-987c-b37ffa925540.jpg)
> 
> * make this prompt appear on the list building page and not in the first page. Otherwise, you just land on the website to discover it and you're ask if you want to install it without having the time to at least check what it is (like the cookies prompt we suffer everywhere)

Alright I will try my best, I never used React before, so there might be some better way of doing it.

Cerceis commented:
![スクリーンショット 2022-09-01 8 54 50](https://user-images.githubusercontent.com/48720021/187804730-d9236353-2f20-4621-9f04-3c88bb810ffe.png)

1. Style changed
2. Added to the global wrapper instead only to the list page, was looking for that, I only found it today. So it prompt the user throughout the entire app, and persist even if they change pages.

BaptisteD commented:
1. Thanks, that's more consistent with the rest of the app's style. Could you put the buttons in blue plz ? We don't use this green/red style inside the app. 

2. Just to be sure that I understood, in which screen this alert pops out ?

Cerceis commented:
> * Thanks, that's more consistent with the rest of the app's style. Could you put the buttons in blue plz ? We don't use this green/red style inside the app.
> * Just to be sure that I understood, in which screen this alert pops out ?

The UX philosophy was to grab user's attention. And to guide which button to click. But anyway I have changed the color back to primary.

Since it's wrapped along in the app wrapper. The pop is going to appear on the **Entire app, which means every single pages.**

BaptisteD commented:
Thanks. 

Is it possible to delay the appearance of this until the actual list building screen ? 
As I said, to avoid jumping to the user's face before he even discovered what the app is.

Cerceis commented:
> Thanks.
> 
> Is it possible to delay the appearance of this until the actual list building screen ? As I said, to avoid jumping to the user's face before he even discovered what the app is.

Well could simply add a 5 second delay until the prompt is initiated if that’s what you want.

BaptisteD commented:
That would be one way to do it. 
However, my question was "is it possible to make it appears on the list building screen".  I can understand a no, but please, answer this specific question.  ^^

Cerceis commented:
> That would be one way to do it.
> However, my question was "is it possible to make it appears on the list building screen". I can understand a no, but please, answer this specific question. ^^

You did not specify, your question are confusing and inconsistent.
Is your question

1) To show the pop up ONLY in the list page
2) Is it possible to make it appear in the list page ( Which prior was only in the index, was then fixed to appear globally)
3) All other page but not the index page.

BaptisteD commented:
So, just so we are sure to talk about the same screen, I'm talking about this one. 
![Screenshot_20220903-101834](https://user-images.githubusercontent.com/2566847/188262461-bdc2478f-ebc0-40f4-8139-c39dc4cc7606.jpg)

So, the question is : is it possible to wait until the user reach this screen before showing up the alert ? 

As a dialog should disable all app functionality that are behind it, it shouldn't be possible to change page when it's open.
You can see this behavior in the "Open a list" screen when you try to delete a list.

AdamLay commented:
Thanks for this @Cerceis - good progress! Yes agree with @BaptisteD , would be good I think if we can wait until the `/list` or `list.tsx` page only (remove from `_app.js` global page). I think users are landing on the app for the first time, dismissing it, then coming back later to ask about the app so many once they've had some time to process it it'll be better to prompt on that page.

Also I'm assuming we can add an "Install App" button into the main menu, if it's not currently running as a PWA - I'm happy to look at this one separately, don't mind. Thanks!

AdamLay commented:
Also can the `package-lock.json` and `yarn.lock` changes be removed from the PR please! Thanks,

Cerceis commented:
> So, just so we are sure to talk about the same screen, I'm talking about this one. ![Screenshot_20220903-101834](https://user-images.githubusercontent.com/2566847/188262461-bdc2478f-ebc0-40f4-8139-c39dc4cc7606.jpg)
> 
> So, the question is : is it possible to wait until the user reach this screen before showing up the alert ?
> 
> As a dialog should disable all app functionality that are behind it, it shouldn't be possible to change page when it's open. You can see this behavior in the "Open a list" screen when you try to delete a list.

Yes, it's possible, but you didn't answer completely. Do you want it ONLY appear on this page, or skip few of the pages prior. This is no a dialog. This is a popup, it does not disable user interaction behind it. Or are you simply implying this should be changed into a dialog rather than a popup? I would totally disagree with that approach. As this feature is not a necessary step that explicitly requires user's immediate action. It would be totally annoying to the user-mid traversing the app if we choose to completely strip user's ability to interact.

Cerceis commented:
> Thanks for this @Cerceis - good progress! Yes agree with @BaptisteD , would be good I think if we can wait until the `/list` or `list.tsx` page only (remove from `_app.js` global page). I think users are landing on the app for the first time, dismissing it, then coming back later to ask about the app so many once they've had some time to process it it'll be better to prompt on that page.
> 
> Also I'm assuming we can add an "Install App" button into the main menu, if it's not currently running as a PWA - I'm happy to look at this one separately, don't mind. Thanks!

I'll work on it. There's a weird phenomenon that cause window.events not being trigger on the list page in the useEffect hook. Trying some workaround for that.

Cerceis commented:
Updated to show the popup only in the list page, but due to the behavior of the event listener, it will still need to be initiated on app load. Hence it will still be included in _app.js. I have also added a whitelist constant that accepts router path, just for future proof incase if in the future other pages are to be added or to change the path to be displayed.

> Also can the package-lock.json and yarn.lock changes be removed from the PR please! Thanks,

I tried removing the yarn.lock but the verbal check returns an error. So I just copied the same yarn from the current dev-branch.

BaptisteD commented:
"As this feature is not a necessary step that explicitly requires user's immediate action. It would be totally annoying to the user-mid traversing the app if we choose to completely strip user's ability to interact."

I completely agree with you here, that's the reason of my request to delay it's appearance. 
But I'm wondering then, if you don't want it to block user interaction, we could use a different pattern than something coming in the middle of the screen. Like a banner 

"A banner displays an important, succinct message, and provides actions for users to address (or dismiss the banner). It requires a user action to be dismissed.

Banners should be displayed at the top of the screen, below a top app bar. They’re persistent and nonmodal, allowing the user to either ignore them or interact with them at any time. Only one banner should be shown at a time."
![unnamed](https://user-images.githubusercontent.com/2566847/188400491-882ff6c1-f74b-4a5b-8efc-26c2dead05da.png)

Cerceis commented:
> "As this feature is not a necessary step that explicitly requires user's immediate action. It would be totally annoying to the user-mid traversing the app if we choose to completely strip user's ability to interact."
> 
> I completely agree with you here, that's the reason of my request to delay it's appearance. But I'm wondering then, if you don't want it to block user interaction, we could use a different pattern than something coming in the middle of the screen. Like a banner
> 
> "A banner displays an important, succinct message, and provides actions for users to address (or dismiss the banner). It requires a user action to be dismissed.
> 
> Banners should be displayed at the top of the screen, below a top app bar. They’re persistent and nonmodal, allowing the user to either ignore them or interact with them at any time. Only one banner should be shown at a time." ![unnamed](https://user-images.githubusercontent.com/2566847/188400491-882ff6c1-f74b-4a5b-8efc-26c2dead05da.png)

Banner works too, but then it will alter the layout of the entire list view, possible breaking while doing so. Also if the user has a very big list. Pushing all the content down is very performance expensive (similar to expansion tabs), lagging the entire app and even possible breaking it. I am already having lags and in some of my older devices by having 1k pt worth of army list (Fire HD 7 2017).

Which is why I choose popup at first. Because it's the "cheapest", and non-breaking way to "plug-in" to the app.
